### PR TITLE
Added @prepare annotation to run custom methods before the module

### PR DESCRIPTION
**Pull request by DavertMik**
Following the question from #4326 I discovered that there was no way to change the module configuration for a specific test (only for groups via GroupObject, a bit complicated way). 

This PR adds a new annotation `@prepare` to Cest and Test classes, which executes a method with a module instance passed into it. This allows preparing the module before the `_before` method is invoked.

```php
/*
 * @prepare disableTransactions
 */
function testDoctrine()
{
}

protected function disableTransactions(Doctrine2 $module)
{
   $module->_reconfigure(['cleanup' => false);
}
```

For instance, this can be used to:

* switch browsers in some acceptance tests
* disable transactions for some Doctrine2, Laravel, Yii2 tests
* start custom browser session for WebDriver (maybe in cloud service), esp with `start: false` option
* ...??? something else to think about.

This PR also makes similar result to be achieved with GroupObject:

```php
class Chrome extends \Codeception\GroupObject
{
    public static $group = 'chrome';

    public function _before()
    {
        $this->getModule('WebDriver')->_reconfigure(['browser' => 'chrome']);
    }
}
```

Also: 

* Refactored Metadata class to accept all possible annotations and store them in $params
* Created new Subscriber to invoke methods in @prepare
* Updated WebDriver to update capabilities on _reconfigure
* Changed Subscriber\Module to reset module config after the test, so it could be updated before
* Simplified Parser

---

Work In Progress:

* [x] tests
* [x] documentation

### Update

* renamed `@init` to `@prepare`
* refactored to use DI to auto-inject modules, helpers, and support classes
* simplified Subscriber

### CODE REVIEW

**Comment by DavertMik** [comment]
Following conversation from #4326 
@ThomasLandauer 

> Two further ideas about that: Allow enabling/disabling of entire modules too.

No. Just no. It makes a very complicated setup then.

> Allow reconfigure of codeception.yml too.

It is out of the scope of a single test. 

>  What about noTransactions(Doctrine) instead of Doctrine noTransactions? This would show clearer that Doctrine is being passed as parameter to the noTransactions function. Plus it would make things possible (if not now, then maybe in the future) like: changeWhatever(Doctrine, Symfony, WebDriver).

Makes sense. I think I can even use autowriring to inject all kind of dependencies with DI!

> And finally I must say that I don't like annotations. I guess the con's are well-known (no syntax highlighting; looks like a comment; hard to handle if you have more than 1 sort of annotations

Yes, this can turn into a long discussion but to my opinion annotations is evil in production code but they fit very good for tests. They provide do what they intend to: they provide meta information for tests. This make test focused on the problem and not on implementation. However, it is good to keep this meta information inside a code as it highly connected with a test itself.

**Comment by ThomasLandauer** [comment]
+1

**Comment by Naktibalda** [comment]
I think that Travis push build is failing because @DavertMik forgot to pull before creating a new branch.

**Comment by DavertMik** [comment]
Yep, that's right. I merged and added documentation

File: docs/06-ModulesAndHelpers.md [hunk 1]
```diff
@@ -264,46 +264,6 @@ In this helper you can replace the parent's methods with your own implementation
 You can also replace the `_before` and `_after` hooks, which might be an option
 when you need to customize starting and stopping of a testing session.
 
-If some of the methods of the parent class should not be used in a child module, you can disable them.
```
Comment (DavertMik) [inline docs/06-ModulesAndHelpers.md]: This part is a bit magical and overcomplicated. I don't think it should be used anymore and probably we will drop this functionality in future releases.

**Pull request merged**